Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm

```diff
@@ -1833,3 +1834,146 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    [handler setStartURLSchemeTaskHandler:^(WKWebView *, id<WKURLSchemeTask> task) {
+        NSString *path = task.request.URL.path;
+        NSString *type = nil;
+        NSString *result = nil;
+        if ([path hasSuffix:@"index.html"]) {
+            result = [NSString stringWithFormat:@"<!DOCTYPE html>"
+                "<script>"
+                "   window.results = [];"
+                "   window.addEventListener('message', function(event) {"
+                "       window.results.push(event.data);"
+                "   });"
+                "</script>"
+                "<script src='test:///script.js'></script>"
+                "<iframe src='http://localhost:%d/subframe.html'></iframe>", port];
+            type = @"text/html";
+        } else if ([path hasSuffix:@"script.js"]) {
+            result = @"window.results.push(window.location.href);";
+            type = @"text/javascript";
+        } else if ([path hasSuffix:@"script_subframe.js"]) {
+            result = @"window.parent.postMessage(window.location.href, '*');";
+            type = @"text/javascript";
+        }
+
+        if (!result) {
+            [task didFailWithError:[NSError errorWithDomain:@"TestWebKitAPI" code:1 userInfo:nil]];
+            return;
+        }
+
+        auto response = adoptNS([[NSURLResponse alloc] initWithURL:task.request.URL MIMEType:type expectedContentLength:[result length] textEncodingName:nil]);
+        [task didReceiveResponse:response.get()];
+        [task didReceiveData:[result dataUsingEncoding:NSUTF8StringEncoding]];
+        [task didFinish];
+    }];
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    [configuration setURLSchemeHandler:handler.get() forURLScheme:@"test"];
+
+    doneCompiling = false;
+    NSString* contentBlocker = @"[{\"action\":{\"type\":\"block\"},\"trigger\":{\"url-filter\":\".*\",\"resource-type\":[\"script\"]}}]";
+    [[WKContentRuleListStore defaultStore] compileContentRuleListForIdentifier:@"WebsitePoliciesTest" encodedContentRuleList:contentBlocker completionHandler:^(WKContentRuleList *list, NSError *error) {
+        EXPECT_TRUE(error == nil);
+        [[configuration userContentController] addContentRuleList:list];
+        doneCompiling = true;
+    }];
+    TestWebKitAPI::Util::run(&doneCompiling);
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 800, 600) configuration:configuration.get()]);
+    [webView setNavigationDelegate:navigationDelegate.get()];
+
+    auto request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"test:///index.html"]];
+    [webView loadRequest:request];
+    [navigationDelegate waitForDidFinishNavigation];
+
+    NSArray<NSString *> *results = [webView objectByEvaluatingJavaScript:@"window.results"];
+
+    EXPECT_EQ(1U, results.count);
+    EXPECT_WK_STREQ("test:///index.html", results[0]);
+
+    [[WKContentRuleListStore defaultStore] _removeAllContentRuleLists];
+}
```

Comment:
Could you also add a test verifying that if you load an http page with an http iframe, or a custom scheme page with a custom scheme iframe, the main frame policies still apply everywhere?

Reply:
Sure, added.